### PR TITLE
docs(aio): Update Universal/SSR guide to align with current CLI

### PR DESCRIPTION
**Pull request by fmalcher**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[x] Documentation content changes
[ ] angular.io application / infrastructure changes
[ ] Other... Please describe:
```

## What is the new behavior?

The current version of the Angular CLI (6.0.8) puts build artifacts in a separate folder under `dist`, names by the app/project name, e.g. `dist/my-app`.
This shuld also be reflected in the docs for people who just follow the guide to add the server-side part to their project.

This PR also adds `webpack-cli` as dependency.
It is needed, it is even included in the example app from the TOH. However, a default Angular CLI project does not contain `webpack-cli`.

To let those changes apply to the docs we might also need to update the TOH example to follow the current behaviour. Please let's discuss this to not break things for users.

## Does this PR introduce a breaking change?
```
[ ] Yes
[ ] No
[x] Maybe? See above.
```

## Other information

This PR also includes the commits from https://github.com/angular/angular/pull/24373 because I branched from there.

### CODE REVIEW

**Comment by fmalcher** [comment]
What I forgot is the universal schematics. They should definitely be mentioned in the docs.
I think good is to use the schematics first and THEN explain the single steps as we would do them manually.

```
ng g universal --clientProject=angular.io-example
```

**Comment by brandonroberts** [comment]
@fmalcher what's remaining on this PR?

**Comment by fmalcher** [comment]
I didn't continue on this because I was not sure whether we should introduce the universal schematics here or not. We would have to restructure the whole thing then.

However, this particular PR can be merged as it brings some sensible changes.
Including the schematics could be a second task then, maybe we could chat about this somehow.
@brandonroberts

File: aio/content/guide/universal.md [hunk 1]
```diff
@@ -369,24 +371,21 @@ which then forwards it to the client in the HTTP response.
 The web server must distinguish _app page requests_ from other kinds of requests.
 
 It's not as simple as intercepting a request to the root address `/`.
-The browser could ask for one of the application routes such as `/dashboard`, `/heroes`, or `/detail:12`.
-In fact, if the app were _only_ rendered by the server, _every_ app link clicked would arrive at the server
+The browser could ask for one of the application routes such as `/dashboard`, `/heroes`, or `/detail/12`.
+In fact, if the app was _only_ rendered by the server, _every_ app link clicked would arrive at the server
```
Comment (gkalpak) [inline aio/content/guide/universal.md]: "were" was correct here: https://www.grammarly.com/blog/was-vs-were/

File: aio/content/guide/universal.md [hunk 2]
```diff
@@ -369,24 +371,21 @@ which then forwards it to the client in the HTTP response.
...
 Data requests also lack extensions but they're easy to recognize because they always begin with `/api`.
-
 All static asset requests have a file extension (e.g., `main.js` or `/node_modules/zone.js/dist/zone.js`).
-
 So we can easily recognize the three types of requests and handle them differently.
 
-1. data request -  request URL that begins `/api`
-2. app navigation - request URL with no file extension
-3. static asset - all other requests.
+1. **Data request:** request URL that begins `/api`
```
Comment (gkalpak) [inline aio/content/guide/universal.md]: While you're at it: _begins_ --> _begins with_

File: aio/content/guide/universal.md [hunk 3]
```diff
@@ -543,13 +545,13 @@ Node server listening on http://localhost:4000
...
+You should see the familiar _Tour of Heroes_ dashboard page.
 
-Navigation via `routerLinks` works correctly.
+Navigation via `routerLink` works correctly.
 You can go from the Dashboard to the Heroes page and back.
-You can click on a hero on the Dashboard page to display its Details page.
+You can also click on a hero on the Dashboard page to display its Details page.
 
-But clicks, mouse-moves, and keyboard entries are inert.
+However, clicks, mouse-moves and keyboard entries are inert:
```
Comment (gkalpak) [inline aio/content/guide/universal.md]: Actually, if you have built with `npm run build:ssr` (as suggested by the tutorial), the `browser` part will also have been built, and thus the page will be interactive (once it is loaded). This part of the guide is inaccurate.

**Review by gkalpak — CHANGES_REQUESTED** [review]

**Comment by fmalcher** [comment]
Thanks @gkalpak for the feedback. I actually like the idea to produce a static-only page first and then add the browser part later.

It seems like the whole guide needed some rework.
In general, we should agree on what the guide wants to tell:

- **1.) A fully-featured solution with everything included (that's the way it is atm)**
  - setting up everything manually from scratch
  - webpack build for the server part
  - NPM scripts
  - combined API/Universal server
- **2.) A simple way that makes it easy to learn the concepts of SSR/Universal without unnecessary overhead**
  - Using `ng generate universal` to set up
  - no webpack build for the server, just `ts-node server` to run (and a red message that this is not suitable for production)
  - no combined API/Universal server but just the Universal part to keep things simple


I actually prefer the second way because it focuses on the necessary party of Universal/SSR. The journey would be like this then:
- Have a simple client-side app
- `ng g universal --clientProject=toh`
- See what has changed
  - `@angular/platform-server` installed
  - `BrowserModule.withServerTransition()` added
  - new `AppServerModule` and tsconfig (entry point for the server build)
  - new part in `angular.json` for the server build
- build both client and server part
  - Server: `ng run toh:server`
  - Client: `ng build --prod`
- See what files have been created in the dist folder
- Introduce the server with its two main duties:
  - rendering the application and delivering the resulting HTML
  - delivering the static assets (i.e. the bundles)
- set up the simple node/express server with just one route for the Universal part:
  - `app.get('*', (req, res) => { res.render('index', { req }); });`
- start the server with `ts-node server`
- Oops, see that the app works but is real static because no JS is involved
  - also see that the bundles are not delivered correctly from the server
- add the server route for static assets
- :tada:

This only features the hard necessary parts of SSR without any overhead for the build setup or anything. What do you think?

File: aio/content/guide/universal.md (hunk 2 above)
Comment (fmalcher) [reply aio/content/guide/universal.md]: resolved

File: aio/content/guide/universal.md (hunk 1 above)
Comment (fmalcher) [reply aio/content/guide/universal.md]: thanks, resolved

**Comment by gkalpak** [comment]
Other than the "webpack build for the server part" and "NPM scripts", I am not sure what other differences there are. For example what does "combined API/Universal server" refer to? (The route that is supposed to handle API requests - but actually does nothing, because of `angular/in-memory-web-api`?)

In general, I like your suggested approach, with two exceptions:

1. Defining a couple of NPM scripts for convenience is a good thing imo (and has really minimal overhead).
2. Showing how to build the server script with webpack (e.g. as a last step) is also helpful imo (although borderline outside the scope of the guide, I must admit). I am on the fence about this one :grin:

But other than that, I totally agree with your suggested plan.

##
@brandonroberts, @jenniferfell: WDYT?

**Comment by fmalcher** [comment]
@gkalpak  I'm with you ;-)
NPM scripts are okay for me as long as we just put them in as an addition.
My problem with the guide when I first read it was that it wanted me to add so much stuff without understanding what I'm doing. I just copy-pasted everything.
That's the point where "less is more" and I like to avoid things such as webpack config and NPM scripts.

What do you think about splitting the guide into two parts?
1.) Fast success with just the necessary explanation
2.) More detailed guide with the missing parts
  - what do all those commands mean
  - NPM scripts
  - How to set up a webpack build for the server

**Comment by fmalcher** [comment]
> For example what does "combined API/Universal server" refer to? (The route that is supposed to handle API requests - but actually does nothing, because of angular/in-memory-web-api?)

Yep, I mixed this up with another thing I saw. However, for the baasic example there is no need IMO for that separate route. Once someone has understood how Universal works they can add more stuff for themselves.

**Comment by gkalpak** [comment]
Hm...not sure how I feel about https://github.com/angular/angular/pull/24375#issuecomment-404081485. I liked the suggestion in https://github.com/angular/angular/pull/24375#issuecomment-404081485 mor :smiley:

Splitting the guide in two parts of different "depth" sounds inconsistent with the rest of the guides.

@brandonroberts, @jenniferfell: Do you have thoughts on this?

**Comment by brandonroberts** [comment]
@fmalcher @gkalpak I don't think it should be split into two guides. I like concepts from both. I think it should be a pretty full-featured guide, but updated with the parts that the CLI handles for you if you're using an existing app where you want to add server-side rendering.

@vikerman may have some thoughts on this also.

**Comment by brandonroberts** [comment]
@fmalcher this guide has been reorganized for better structure and flow. Will you check to see if this PR is still relevant?

**Comment by gkalpak** [comment]
There's a PR for the second one: #26865

**Comment by brandonroberts** [comment]
@fmalcher will you open a PR for the first one? Thanks

**Comment by brandonroberts** [comment]
Closing this one as obsolete.

**Pull request closed without merging**